src/rust/src/params.rs

```diff
@@ -0,0 +1,1445 @@
+        }
+    }
+
+    if _present - initial < 2 || _present - initial > 5 {
+        panic!("last language code should be xxx.\n");
+    }
+
+    if _present - initial == 5 {
+        let sub_slice = &lang[initial.._present];
+        if !sub_slice.contains('-') {
+            panic!("last language code is not of the form xxx-xx\n");
+        }
+    }
+}
+
+fn parse_708_services(opts: &mut CcxSOptions, s: &str) {
+    if s.starts_with("all") {
+        let charset = if s.len() > 3 { &s[4..s.len() - 1] } else { "" };
+        opts.settings_dtvcc.enabled = true;
+        opts.enc_cfg.dtvcc_extract = true;
+        opts.enc_cfg.all_services_charset = charset.to_owned();
+        opts.enc_cfg.services_charsets = vec![String::new(); CCX_DTVCC_MAX_SERVICES];
+
+        for i in 0..CCX_DTVCC_MAX_SERVICES {
+            opts.settings_dtvcc.services_enabled[i] = true;
+            opts.enc_cfg.services_enabled[i] = true;
+        }
+
+        opts.settings_dtvcc.active_services_count = CCX_DTVCC_MAX_SERVICES;
+        return;
+    }
+
+    let mut services = Vec::new();
+    let mut charsets = Vec::new();
+    for c in s.split(',') {
+        let mut service = String::new();
+        let mut charset = None;
+        let mut inside_charset = false;
+
+        for e in c.chars() {
+            if e.is_ascii_digit() && !inside_charset {
+                service.push(e);
+            } else if e == '[' {
+                inside_charset = true;
+            } else if e == ']' {
+                inside_charset = false;
+            } else if inside_charset {
+                if charset.is_none() {
+                    charset = Some(String::new());
+                }
+                charset.as_mut().unwrap().push(e);
+            }
+        }
+        if service.is_empty() {
+            continue;
+        }
+        services.push(service);
+        charsets.push(charset.clone());
+    }
+
+    opts.enc_cfg.services_charsets = vec![String::new(); CCX_DTVCC_MAX_SERVICES];
+
+    for (i, service) in services.iter().enumerate() {
+        let svc = service.parse::<usize>().unwrap();
+        if !(1..=CCX_DTVCC_MAX_SERVICES).contains(&svc) {
+            panic!("[CEA-708] Malformed parameter: Invalid service number ({}), valid range is 1-{}.\n", svc, CCX_DTVCC_MAX_SERVICES);
+        }
+        opts.settings_dtvcc.services_enabled[svc - 1] = true;
+        opts.enc_cfg.services_enabled[svc - 1] = true;
+        opts.settings_dtvcc.enabled = true;
+        opts.enc_cfg.dtvcc_extract = true;
+        opts.settings_dtvcc.active_services_count += 1;
+
+        if charsets.len() > i && charsets[i].is_some() {
+            opts.enc_cfg.services_charsets[svc - 1] = charsets[i].as_ref().unwrap().to_owned();
+        }
+    }
+
+    if opts.settings_dtvcc.active_services_count == 0 {
+        panic!("[CEA-708] Malformed parameter: no services\n");
+    }
+}
+
+pub fn parse_parameters(opt: &mut CcxSOptions, args: &Args, tlt_config: &mut CcxSTeletextConfig) {
```

Comment:
Functions that take a `opt: &mut CcxSOptions`, should rather be implemented on `CcxSOptions` using `impl CcxSOptions`

Reply:
Side note: Not sure why these structs were named with `_s_` in them in C. You can drop these.
`CcxSOptions` => `CcxOptions`
`CcxSTeletextConfig` => `CcxTeletextConfig`

Reply:
@PunitLodha now there's a conflict as opt is &mut and tlt_config is also &mut, so how will Implement for both?

Reply:
Most functions take in `opt: &mut CcxOptions`, so you can use `impl CcxOptions` and pass `tlt_config` as `tlt_config: &mut CcxTeletextConfig`

Reply:
Done!